common/djangoapps/util/password_policy_validators.py

```diff
@@ -47,6 +63,92 @@ def password_max_length():
     return max_length
 
 
+def password_complexity():
+    """
+    :return: A dict of complexity requirements from settings
+    """
+    complexity = {}
+    if settings.FEATURES.get('ENFORCE_PASSWORD_POLICY', False):
+        complexity = getattr(settings, 'PASSWORD_COMPLEXITY', {})
+
+    valid_complexity = {x: y for x, y in complexity.iteritems() if x in _allowed_password_complexity}
+
+    if not password_complexity.logged:
+        invalid = frozenset(complexity.keys()) - frozenset(valid_complexity.keys())
+        for key in invalid:
+            log.warning('Unrecognized %s value in PASSWORD_COMPLEXITY setting.', key)
+        password_complexity.logged = True
+
+    return valid_complexity
+
+
+# Declare static variable for the function above, which helps avoid issuing multiple log warnings.
+# We don't instead keep a cached version of the complexity rules, because that might trip up unit tests.
+password_complexity.logged = False
+
+
+def _password_complexity_descriptions(which=None):
+    """
+    which: A list of which complexities to describe, None if you want the configured ones
+    :return: A list of complexity descriptions
+    """
+    descs = []
+    complexity = password_complexity()
+    if which is None:
+        which = complexity.keys()
+
+    for key in _allowed_password_complexity:  # we iterate over allowed keys so that we get the order right
+        value = complexity.get(key, 0) if key in which else 0
+        if not value:
+            continue
+
+        if key == 'ALPHABETIC':
```

Comment:
TIL there's no switch cases in python.

Reply:
Yeah I'm torn on that. On the one hand, switches are real weird. But surely there was room to improve on the syntax inside python? Ah well.